File: presentation_generate/ppt_th/temp_trial.py
from pptx import Presentation
import os
import re

def copy_slide(prs, slide):
    slide_layout = slide.slide_layout  # 使用原幻灯片的布局
    new_slide = prs.slides.add_slide(slide_layout)  # 创建新幻灯片

    # 复制原幻灯片上的所有形状
    for shape in slide.shapes:
        if shape.is_placeholder:
            # 对占位符类型的形状进行复制
            new_shape = new_slide.shapes.add_placeholder(shape.placeholder_format.idx, shape.left, shape.top,
                                                         shape.width, shape.height)
            new_shape.text = shape.text  # 复制文本内容（如果有）
        elif shape.shape_type == 13:  # 图片（图像类型为13）
            # 复制图片
            new_slide.shapes.add_picture(shape.image, shape.left, shape.top, shape.width, shape.height)
        else:
            # 其他形状（如矩形、圆形等）
            new_shape = new_slide.shapes.add_shape(shape.auto_shape_type, shape.left, shape.top, shape.width,
                                                   shape.height)
            new_shape.text = shape.text  # 复制文本内容（如果有）

    return new_slide

def fill_ppt_content(template_route, save_route):

    def match_placeholder(tag, text):
        pattern =  repr(tag) + r"(\d+)"

        # 获取第一个匹配项
        match = re.search(pattern, text.lower())

        if match:
            return int(match.group(1))  # 输出第一个匹配的数字内容
        return None

    # 打开PPT文件
    prs = Presentation(template_route)
    # 遍历所有幻灯片
    xml_slides = prs.slides._sldIdLst  # 获取幻灯片ID列表
    slides = list(xml_slides)

    # 添加标题slide
    new_title_slide = copy_slide(prs, 0)
    for shape in new_title_slide.shapes:
        if not shape.has_text_frame:
            continue
        paragraphs = shape.text_frame.paragraphs
        # for i in range(1, len(paragraphs)):
        #     shape.text_frame.remove_paragraph(i)
        for run in paragraphs[0].runs:
            if "title" in run.text.lower():
                run.text = "大模型的前世今生"
            elif "speaker" in run.text.lower():
                run.text = "JOJO"

    new_index_slide = copy_slide(prs, 1)
    for shape in new_index_slide.shapes:
        if not shape.has_text_frame:
            continue
        paragraphs = shape.text_frame.paragraphs
        # for i in range(1, len(paragraphs)):
        #     shape.text_frame.remove_paragraph(i)
        for run in paragraphs[0].runs:
            if "title" in run.text.lower():
                run.text = "大模型的前世今生"
            elif "speaker" in run.text.lower():
                run.text = "JOJO"

    new_subtitle_slide = copy_slide(prs, 2)
    for shape in new_subtitle_slide.shapes:
        if not shape.has_text_frame:
            continue
        paragraphs = shape.text_frame.paragraphs
        # for i in range(1, len(paragraphs)):
        #     shape.text_frame.remove_paragraph(i)
        for run in paragraphs[0].runs:
            if "title" in run.text.lower():
                run.text = "Transformer介绍"
            elif "order" in run.text.lower():
                run.text = "01"

    new_content_slide = copy_slide(prs, 3)
    content_list = [
        ("优化沟通流程", "项目旨在改善公司内部沟通，提高工作效率。"),
        ("明确项目目标", "通过PPT汇报确保团队成员对目标与成果有共同理解。"),
        ("完成需求分析", "已进行需求分析并完成初步设计，确定项目实施路径。"),
        ("搭建协作平台", "团队协作平台已搭建完成，初步测试反馈良好。"),
        ("采用AI技术", "利用AI辅助设计提高PPT制作效率与质量，集成数据分析工具。")
    ]
    for shape in new_content_slide.shapes:
        if not shape.has_text_frame:
            continue
        paragraphs = shape.text_frame.paragraphs
        # for i in range(1, len(paragraphs)):
        #     shape.text_frame.remove_paragraph(i)
        for run in paragraphs[0].runs:
            subtitle_res = match_placeholder("subtitle", run.text)
            content_res = match_placeholder("content", run.text)
            if subtitle_res:
                run.text = content_list[content_res - 1][0]
            elif "title" in run.text.lower():
                run.text = "1.1 Transformer发展史"
            elif content_res:
                run.text = content_list[content_res - 1][1]

    for slide_id in xml_slides:
        prs.slides._sldIdLst.remove(slide_id)

    # 保存修改后的PPT文件
    prs.save(save_route)



    # 删除原本的slide
    for slide in slides:
        xml_slides.remove(slide)



if __name__ == '__main__':
    # 获取当前文件所在的路径
    dir_path = os.path.dirname(os.path.realpath(__file__))
    template_route = os.path.join(dir_path, "Template", "Design-qwen.pptx")
    save_route = os.path.join(dir_path, "GeneratedPresentations", "test.pptx")
    fill_ppt_content(template_route, save_route)

File: presentation_generate/ppt_th/temp_trial.js
const pptgen = require('pptgenjs');  // 引入PptGenJS库
const fs = require('fs');
const path = require('path');

function copySlide(ppt, slideIndex) {
  // 获取源幻灯片
  const slide = ppt.slides[slideIndex];

  // 创建一个新的幻灯片
  const slideLayout = slide.layout;  // 获取源幻灯片的布局
  const newSlide = ppt.addSlide(slideLayout);

  // 复制源幻灯片的所有形状到新幻灯片
  slide.shapes.forEach(shape => {
    if (shape.isPlaceholder) {
      const newShape = newSlide.addPlaceholder(shape.placeholderFormat.idx, shape.left, shape.top, shape.width, shape.height);
      if (shape.hasTextFrame) {
        newShape.textFrame.clear();  // 清除文本框中的内容
        shape.textFrame.paragraphs.forEach(paragraph => {
          const newParagraph = newShape.textFrame.addParagraph();
          paragraph.runs.forEach(run => {
            const newRun = newParagraph.addRun();
            newRun.text = run.text;
          });
        });
      }
    } else {
      if (shape.hasTextFrame) {
        const newShape = newSlide.addTextbox(shape.left, shape.top, shape.width, shape.height);
        newShape.textFrame.clear();
        shape.textFrame.paragraphs.forEach(paragraph => {
          const newParagraph = newShape.textFrame.addParagraph();
          paragraph.runs.forEach(run => {
            const newRun = newParagraph.addRun();
            newRun.text = run.text;
          });
        });
      } else if (shape.shapeType === 13) {  // 图片
        newSlide.addPicture(shape.image.filename, shape.left, shape.top, shape.width, shape.height);
      }
    }
  });

  return newSlide;
}

function matchPlaceholder(tag, text) {
  const pattern = new RegExp(tag + "(\\d+)", "i");

  // 获取第一个匹配项
  const match = text.match(pattern);
  if (match) {
    return parseInt(match[1], 10);  // 输出第一个匹配的数字内容
  }
  return null;
}

function fillPptContent(templateRoute, saveRoute) {
  const ppt = new pptgen.Presentation();
  ppt.load(fs.readFileSync(templateRoute));

  const contentList = [
    ["优化沟通流程", "项目旨在改善公司内部沟通，提高工作效率。"],
    ["明确项目目标", "通过PPT汇报确保团队成员对目标与成果有共同理解。"],
    ["完成需求分析", "已进行需求分析并完成初步设计，确定项目实施路径。"],
    ["搭建协作平台", "团队协作平台已搭建完成，初步测试反馈良好。"],
    ["采用AI技术", "利用AI辅助设计提高PPT制作效率与质量，集成数据分析工具。"]
  ];

  // 添加标题slide
  let newTitleSlide = copySlide(ppt, 0);
  newTitleSlide.shapes.forEach(shape => {
    if (shape.hasTextFrame) {
      shape.textFrame.paragraphs.forEach(paragraph => {
        paragraph.runs.forEach(run => {
          if (run.text.toLowerCase().includes("title")) {
            run.text = "大模型的前世今生";
          } else if (run.text.toLowerCase().includes("speaker")) {
            run.text = "JOJO";
          }
        });
      });
    }
  });

  let newIndexSlide = copySlide(ppt, 1);
  newIndexSlide.shapes.forEach(shape => {
    if (shape.hasTextFrame) {
      shape.textFrame.paragraphs.forEach(paragraph => {
        paragraph.runs.forEach(run => {
          if (run.text.toLowerCase().includes("title")) {
            run.text = "大模型的前世今生";
          } else if (run.text.toLowerCase().includes("speaker")) {
            run.text = "JOJO";
          }
        });
      });
    }
  });

  let newSubtitleSlide = copySlide(ppt, 2);
  newSubtitleSlide.shapes.forEach(shape => {
    if (shape.hasTextFrame) {
      shape.textFrame.paragraphs.forEach(paragraph => {
        paragraph.runs.forEach(run => {
          if (run.text.toLowerCase().includes("title")) {
            run.text = "Transformer介绍";
          } else if (run.text.toLowerCase().includes("order")) {
            run.text = "01";
          }
        });
      });
    }
  });

  let newContentSlide = copySlide(ppt, 3);
  newContentSlide.shapes.forEach(shape => {
    if (shape.hasTextFrame) {
      shape.textFrame.paragraphs.forEach(paragraph => {
        paragraph.runs.forEach(run => {
          let subtitleRes = matchPlaceholder("subtitle", run.text);
          let contentRes = matchPlaceholder("content", run.text);
          if (subtitleRes) {
            run.text = contentList[contentRes - 1][0];
          } else if (run.text.toLowerCase().includes("title")) {
            run.text = "1.1 Transformer发展史";
          } else if (contentRes) {
            run.text = contentList[contentRes - 1][1];
          }
        });
      });
    }
  });

  ppt.save(fs.createWriteStream(saveRoute));
}

const templateRoute = path.join(__dirname, "Template", "Design-qwen.pptx");
const saveRoute = path.join(__dirname, "GeneratedPresentations", "test.pptx");
fillPptContent(templateRoute, saveRoute);
